src/engine/SCons/Action.py

```diff
@@ -998,6 +1008,44 @@ def get_implicit_deps(self, target, source, env, executor=None):
                     res.append(env.fs.File(d))
         return res
 
+    def _get_implicit_deps_heavyweight(self, target, source, env, executor):
+        # Avoid circular and duplicate dependencies by not providing source,
+        # target, or executor to subst_list. This causes references to
+        # $SOURCES, $TARGETS, and all related variables to disappear.
+        from SCons.Subst import SUBST_SIG
+        cmd_list = env.subst_list(self.cmd_list, SUBST_SIG)
+        res = []
+
+        for cmd_line in cmd_list:
+            if cmd_line:
+                for entry in cmd_line:
+                    d = str(entry)
+                    if not d.startswith(('&', '-', '/') if os.name == 'nt'
+                                        else ('&', '-')):
+                        m = strip_quotes.match(d)
+                        if m:
+                            d = m.group(1)
+
+                        # For now, only match files, not directories.
+                        p = os.path.abspath(d) if os.path.isfile(d) else \
+                            env.WhereIs(d)
+                        if p:
+                            res.append(env.fs.File(p))
+                        else:
+                            # Try to find the dependency in any source
```

Comment:
I think In general including path into repo is deprecated.  Should just look for non-repo file. There's a way to query if File is known without creating the node.

Reply:
So are you saying the “else” case should be removed because repo support is not needed?